src/main/java/seedu/address/model/warehouse/UniqueWarehouseList.java

```diff
@@ -0,0 +1,142 @@
+import seedu.address.model.warehouse.exceptions.WarehouseNotFoundException;
+
+/**
+ * A list of warehouses that enforces uniqueness between its elements and does not allow nulls.
+ * A warehouse is considered unique by comparing using {@code Warehouse#isSameWarehouse(Warehouse)}. As such, adding
+ * and updating of warehouses uses Warehouse#isSameWarehouse(Warehouse) for equality so as to ensure that the person
+ * being added or updated is unique in terms of identity in the UniqueWarehouseList.
+ * However, the removal of a warehouse uses Warehouse#equals(Object) so as to ensure that the warehouse with
+ * exactly the same fields will be removed.
+ *
+ * Supports a minimal set of list operations.
+ *
+ * @see Warehouse#isSameWarehouse(Warehouse)
+ */
+public class UniqueWarehouseList implements Iterable<Warehouse> {
+    private final ObservableList<Warehouse> internalList = FXCollections.observableArrayList();
+    private final ObservableList<Warehouse> internalUnmodifiableList =
+            FXCollections.unmodifiableObservableList(internalList);
+
+    /**
+     * Returns true if the list contains an equivalent warehouses as the given argument.
+     */
+    public boolean contains(Warehouse toCheck) {
+        requireNonNull(toCheck);
+        return internalList.stream().anyMatch(toCheck::isSameWarehouse);
+    }
+
+    /**
+     * Adds a warehouse to the list.
+     * The warehouse must not already exist in the list.
+     */
+    public void add(Warehouse warehouseToAdd) {
+        requireNonNull(warehouseToAdd);
+        if (contains(warehouseToAdd)) {
+            throw new DuplicateWarehouseException();
+        }
+        internalList.add(warehouseToAdd);
+    }
+
+    /**
+     * Replaces the warehouse {@code target} in the list with {@code editedWarehouse}.
+     * {@code target} must exist in the list.
+     * The warehouse identity of {@code editedWarehouse} must not be the same as another existing warehouse in the list.
+     */
+    public void setWarehouse(Warehouse target, Warehouse editedWarehouse) {
+        requireAllNonNull(target, editedWarehouse);
+
+        int index = internalList.indexOf(target);
+        if (index == -1) {
+            throw new PersonNotFoundException();
+        }
+
+        if (!target.isSameWarehouse(editedWarehouse) && contains(editedWarehouse)) {
+            throw new DuplicatePersonException();
+        }
+
+        internalList.set(index, editedWarehouse);
+    }
+
+    /**
+     * Removes the equivalent warehouse from the list.
+     * The warehouse must exist in the list.
+     */
+    public void remove(Warehouse toRemove) {
+        requireNonNull(toRemove);
+        if (!internalList.remove(toRemove)) {
+            throw new WarehouseNotFoundException();
+        }
+    }
+
+    /**
+     * Replaces the current list of warehouses in {@code internalList} with a new {@code replacement}.
+     */
+    public void setWarehouses(UniqueWarehouseList replacement) {
```

Comment:
Do you think replaceWarehouseList would be clearer here since the above method to change one warehouse differ only by a single letter? :)